firmware/defmt-rtt/src/lib.rs

```diff
@@ -27,6 +27,7 @@ struct Logger;
 
 static TAKEN: AtomicBool = AtomicBool::new(false);
 static INTERRUPTS_ACTIVE: AtomicBool = AtomicBool::new(false);
+static mut ENCODER: defmt::Encoder = defmt::Encoder::new();
```

Comment:
I think we should wrap this in a `RacyCell` to avoid potential UB. As it is, I think LLVM will mark this with `noalias` (?) metadata which may not be the right thing for a shared global. I believe all Atomic* types internally use UnsafeCells to avoid issues like that (and they still have unsolved UB issues like `rust-lang/rust 55005`; though that is due to the "deferenceable" attribute). cc @jonas-schievink to sanity check my claim here

I don't think there's a `RacyCell` on crates.io but it basically is a `Sync` UnsafeCell:
``` rust
pub struct RacyCell<T>(core::cell::UnsafeCell<T>);

// this type has no synchronization; synchronization must be enforced manually
unsafe impl<T> Sync for RacyCell<T> {}

impl<T> RacyCell<T> {
    pub const fn new(val: T) -> Self {
        Self(core::cell::UnsafeCell::new(val))
    }

    pub fn get(&self) -> *mut T {
        self.0.get()
    }
}

static UNSYNCHRONIZED_THING: RacyCell<NotPlainOldDataType> =
    RacyCell::new(NotPlainOldDataType::new());

fn usage() {
    // disable interrupts
    unsafe { (*UNSYNCHRONIZED_THING.get()).mutating_method() }
    // enable interrupts
}
```

Reply:
I don't think this is necessary. `UnsafeCell` only enables interior mutability, but `static mut` is already mutable. It just needs synchronization to ensure no 2 `&mut` to it ever exist at the same time, but `UnsafeCell` also needs that.

Reply:
> It just needs synchronization to ensure no 2 &mut to it ever exist at the same time, but UnsafeCell also needs that.

I agree with the Rust side of things; I was more worried with the less visible LLVM codegen side of things but I checked and see that `rustc` is not emitting `noalias` (ATM at least) and that `dereferenceable` is applied to _all_ mutable references regardless of whether they come from `UnsafeCell` or not so sticking an `UnsafeCell` into a `static` vs using a plain `static mut` produces the same LLVM IR.

Never mind me then. I'm OK with keeping this (`static mut` version) as it is.

Reply:
`noalias` is not applied to `&mut` references due to LLVM bugs, it will be applied once those are fixed. But from looking at the code it seems that even that wouldn't cause problems, since it never creates conflicting references to the static.